### DISCUSSION TITLE
how the vertex shader and fragment shader work?

### DISCUSSION THREAD

**Discussion by TopGun34** [mkkellogg/GaussianSplats3D#235] [Q&A] [answered]
there are two set of shaders in SplatMesh.js
1. vertexShaderSource + fragmentShaderSource in **buildMaterial** : which is used to render Gaussians, we treat each Gaussian as a instance of four points, in VS transform each point based the  Gaussian 2D projection, the four point are similar to AABB of this Gaussian. After rasterization,  we can get the pixel's color covered by this Gaussian in FS.
2. vsSource + fsSource in **setupDistancesComputationTransformFeedback**: which is used for transformfeedback to calculate the distance

For the second one, there has code as below to config and use it:
**const vertexShader = createShader(gl, gl.VERTEX_SHADER, vsSource);
const fragmentShader = createShader(gl, gl.FRAGMENT_SHADER, fsSource);
gl.drawArrays(gl.POINTS, 0, this.getSplatCount());**

But for the first one, i haven't find similar code , i think we should have some code as below to draw the quads which cover each Gaussian:
**gl.drawArraysInstanced(gl.TRIANGLE_FAN, 0, 4, vertexCount);**
 
Anyone can help me to understand the code?

**Comment by mkkellogg** [comment, marked as answer]
For the first one, which is the shader for rendering the splats, the actual calls to draw the splat geometry are in the Three.js library. I try to use Three.js where ever I can in this project and for the rendering of the splats, all I need to provide to Three.js is the shader code and some setup for the splat quad geometry (which is in the `buildGeometry()` function). Three.js will internally make a call to `gl.drawArraysInstanced()`. However, Three.js doesn't have built-in support for using transform feedback functionality, so I had to write all of it myself in raw WebGL; that is why you see the call to functions like `gl.drawArrays()` for that part.

**Reply by TopGun34** [reply]
thank you so much, it's clear to me now
